### ISSUE TITLE
Change Attribute Type

### ISSUE THREAD

**Issue by objecttothis** [opensourcepos/opensourcepos#2317]
### Background information

__IMPORTANT: If you choose to ignore this issue report template, your issue will be closed as we cannot help without the requested information.__

Please make sure you tick (add an x between the square brackets with no spaces) the following check boxes:

- [x] Reporting an issue of an unmodified OSPOS installation
- [x] Checked [open and closed issues](https://github.com/opensourcepos/opensourcepos/issues?utf8=%E2%9C%93&q=is%3Aissue) and no similar issue was already reported (please make sure you searched!)
- [x] Read [README](https://github.com/opensourcepos/opensourcepos/blob/master/README.md), [WHATS_NEW](https://github.com/opensourcepos/opensourcepos/blob/master/WHATS_NEW.txt), [INSTALL.md](https://github.com/opensourcepos/opensourcepos/blob/master/INSTALL.md) and [UPGRADE](https://github.com/opensourcepos/opensourcepos/blob/master/UPGRADE.txt)
- [x] Read the [FAQ](https://github.com/opensourcepos/opensourcepos#faq) for any known install and/or upgrade gotchas (in specific PHP extensions installed)
- [x] Read the [wiki](https://github.com/opensourcepos/opensourcepos/wiki)
- [x] Executed any database upgrade scripts if an upgrade pre 3.0.0 (e.g. database/2.4_to_3.0.sql)
- [x] Aware the installation code is in [bintray](https://bintray.com/jekkos/opensourcepos/opensourcepos/view/files?sort=updated&order=asc#files) (see README), and [GitHub master](https://github.com/opensourcepos/opensourcepos/tree/master) is for [developers only](https://github.com/opensourcepos/opensourcepos/wiki/Development-setup) and therefore not complete nor stable

### Installation information

- OSPOS version is: 3.3.0
- OSPOS git commit hash is: 4f5ad57
- PHP version is: 5.6.38
- MySQL or MariaDB version is: 5.7.24
- OS and version is: FreeBSD 11.2
- WebServer is: Apache 2.4
- Selected language is: English
- (If applicable) Docker installation: N/A
- (If applicable) Installation package for the LAMP/LEMP stack is: N/A

### New Feature

Currently after the migration and after creating an attribute the option to change the type of attribute is unavailable.  The ability to "migrate" to a different attribute type is especially important after the migration to 3.3.0 but could also be helpful for user created attributes.

The change would require a confirmation, because of the possibility of causing problems with the data and after accepting should kick off an appropriate migration script to migrate the database field from one type to the other.  

Assumptions: 
- The data can be readily converted in the database.  This applies most if migrating to the DATETIME attribute since the formatting must be correct to change the database field to datetime.  Perhaps give a downloadable csv of all items that do not have valid data in that attribute for the new attribute type and not allow proceeding until all of the data has been fixed.  I recommend building a report for this and when the attribute type dropdown gets changed to change the button text from submit to "audit".  Then if the audit passes, change the button back to submit to move forward with the migration.

The following migrations are easy:
- From DATETIME or DROPDOWN to TEXT.  I don't understand what the GROUP attribute type is for, so I don't know if it's easy or not.
- From TEXT to DROPDOWN should be straight forward since the database field type won't change.

The following migrations are potentially more complicated:
- TEXT to DATETIME should be reasonable as long as the data is formatted properly in the database.

The following migrations should not be allowed:
- GROUP to anything else.
- DROPDOWN to DATETIME.  I could be convinced otherwise but it seems problematic.  Perhaps its no more complicated than TEXT to DATETIME in which case we should allow it.

Thoughts?

**Comment by objecttothis** [comment]
As proof of concept, I manually changed the Attribute type for one of my pre-3.3.0 migrated attributes from TEXT to DROPDOWN in the attribute_definitions and then in the attribute definition view I added all the acceptable values for the dropdown.  In the items table it shows the attribute properly, but in the edit item view it always shows the attribute as the first possible option rather than properly pulling the value from the database.  So at least TEXT to DROPDOWN should just be minimal adjustments to allow the change from the edit attribute view and fix what is causing the edit item view to not display the current "selected" attribute value properly.

**Comment by objecttothis** [comment]
![annotation 2019-01-24 124038](https://user-images.githubusercontent.com/17935339/51665682-c8de0780-1fd5-11e9-9c05-18fb084042ac.jpg)

**Comment by objecttothis** [comment]
I don't think we should hold up 3.3.0 for this feature, but I'm guessing that people will ask for it.  Since I need the feature for my operation, I don't mind doing the legwork on it.

**Comment by objecttothis** [comment]
Here is the data showing the assigned attributes, their values and counts.  For example, during the import the TEXT attribute location with value Р (attribute_id 8704) has 587 rows with different item_id but then when I selected Р from the dropdown of "converted" Location attribute it doesn't use 8704 but creates a new attribute_id.
![data analysis](https://user-images.githubusercontent.com/17935339/51670743-bddca480-1fe0-11e9-81bb-6da4fff8a7cb.jpg)

@jekkos @daN4cat  ideas?

**Comment by objecttothis** [comment]
@jekkos, so this is an issue caused by the migration script?  If I understand correctly, the attribute value table should have only one row with the same attribute value and assign to it a unique attribute_id? Then the attribute_links would have in each row an item_id with an associated definition_id and attribute_id.  I'm  going to play around with it, but what I think is happening is that after I manually change the attribute type to dropdown and create the values, the software isn't checking to see if the value already exists but instead creating a new attribute_id and assigning what happens to be the same value.  If this is true then we don't need to modify the migration script, but instead modify the code that creates a dropdown value to first check to make sure that the value doesn't already exist before adding a row in attribute_values.

**Comment by objecttothis** [comment]
OK, I confirmed it.  This is a bug with the way that a dropdown value gets created.  When I did a manual UPDATE query in the database and set the attribute id that was created by the dropdown "add value" to the same attribute_id as the pre-existing attribute_id from the migration, and therefore it was unique, it works just fine.  Meaning, when you enter the item edit view it actually shows the current value as selected.  This means that we just need to add a check into the function that creates a value in the attribute view to see if the value already exists, and use that attribute_id if it does rather than creating a new attribute_id.  

@jekkos attributes has been your baby.  Do you want to make this change or do you want me to.  I can do it, but I need to get to know how you are doing things in the code a little better.  Just let me know.

**Comment by objecttothis** [comment]
The ability to change the attribute type, I'm glad to implement unless you want to do it.  I'm just asking about the check in the dropdown value creation.

**Comment by jekkos** [comment]
Ok just wanted to say that indeed the dropdown values are created a bit differentl 

* To show a value in the dropdown you need to have an entry in attribute_links without item_id reference. If this one is present it will be shown in the select box
* The selected values always refer to the same attribute value (using its attribute_id) and also refer to an item_id

so you need both these two to have a correct migration.

**Comment by jekkos** [comment]
Feel free to propose an enhancement for the migration. But as you said before, maybe we won't allow any from/to combination. From text to dropdown or the other way might be enough here.

**Comment by jekkos** [comment]
no not working on it ..

**Comment by objecttothis** [comment]
OK, super.  I'll work on it and submit the code once I'm done testing.

**Comment by objecttothis** [comment]
Since pull request #2325 is ready for testing before merging with the master I'm going to start on the code to enable the dropdown to change the type of attribute and the associated logic.  The dropdown should allow the following scenarios.   Someone please confirm:

TEXT -> DROPDOWN (alters attribute_definition table)
TEXT -> DATETIME (Assumes that all data is properly formatted for DATETIME. Alters attribute definition, then alters attribute values table and populates attribute_datetime with the contents of attribute_value then replaces attribute_value and that row with NULL)
DROPDOWN -> TEXT (alters attribute_definition table and then drops all attribute links that have that definition_id AND item_id = NULL)

**Comment by objecttothis** [comment]
@opensourcepos/org we need to decide how the TEXT -> DATETIME should behave when there is invalid data.  Right now, my code needs to copy the attribute_values.attribute_value column to the attribute_values.attribute_datetime column and then set the attribute_values.attribute_value column to NULL.

As far as I can tell, we have three options:
- abort the conversion with a message saying that not all the data is in the correct format.
- do the conversion and ignore incorrectly formatted data then set the attribute_value column to NULL for that definition.
- Do the conversion and ignore incorrectly formatted data but not set the attribute_value column to NULL so as to not lose data.

- The safest route would be to abort with a message, but without a specific list of which conversions are incorrectly formatted, the user is left guessing which attribute_values are causing the problem.

In addition, if for some reason an attribute_value that is being converted happens to also be used by a different definition_id, we wouldn't want to set attribute_value to null during the conversion.  This would be a rare case, but is theoretically possible.

From what I can see, there isn't anything in the code that would break if both the attribute_value and the attribute_datetime are populated.  Thoughts?

**Comment by objecttothis** [comment]
I posted this in gitter but no one seems to be using it, so I don't think it was seen.

Anyone versed in CI Query builder and want to help me figure out how to get this right? I'm implementing the conversion from TEXT to DATETIME. The code needs to be able to execute the following SQL:
UPDATE ospos_attribute_values
INNER JOIN ospos_attribute_links
ON ospos_attribute_links.attribute_id = ospos_attribute_values.attribute_id
SET ospos_attribute_values.ospos_attribute_datetime = ospos_attribute_values.ospos_attribute_value
SET ospos_attribute_values.ospos_attribute_value = NULL
WHERE ospos_attribute_links.definition_id = 8;
My PHP is as follows:
             $this->db->join('ospos_attribute_links','ospos_attribute_values.attribute_id = ospos_attribute_links.attribute_id','inner');
            $this->db->set('ospos_attribute_values.attribute_datetime','ospos_attribute_values.attribute_value');
            $this->db->set('ospos_attribute_values.attribute_value', NULL);
            $this->db->where('ospos_attribute_links.definition_id',$definition_id);
            $success = $this->db->update('ospos_attribute_values','ospos_attribute_links');
The problem is that the query that it is generating (from the log) looks like this:
UPDATE ospos_attribute_values SET ospos_attribute_values.attribute_datetime = 'ospos_attribute_values.attribute_value', ospos_attribute_values.attribute_value = NULL, ospos_attribute_links = ''
WHERE ospos_attribute_links.definition_id = '8'
It's completely ignoring the join() function call.
So, of course it errors out, because without the JOIN it doesn't know what ospos_attribute_links.defition_id is.
Any ideas?

**Comment by jekkos** [comment]
Not sure about the set() call. Most of the time we use an associative array with update() to update the values at hand. Can you try this, so put the field names as keys and then values as value in the array?
 Then check what query it generates

**Comment by objecttothis** [comment]
The problem, I think, is that there is more than one table involved.  I think what you are suggesting is what we are doing in the save_link() function of the Attribute model:

`
		if($this->link_exists($item_id, $definition_id))
		{
			$this->db->where('definition_id', $definition_id);
			$this->db->where('item_id', $item_id);
			$this->db->where('sale_id');
			$this->db->where('receiving_id');
			$this->db->update('attribute_links', array('attribute_id' => $attribute_id));
		}
`
The problem is that this only involves the attribute_links table.  I'm needing a solution that takes all rows from the attribute_links table where the definition id equals the one the user is changing and combines them with the value from attribute_values.  The two tables are connected by the attribute_id.  How would I do what you are suggesting with an effective join?

It seems that the official stance from CI is that you can't use $this->$db->join() with $this->$db->update().  I found two separate solutions online (https://stackoverflow.com/questions/42323649/update-and-join-query-together-in-codeigniter and https://developer-paradize.blogspot.com/2013/06/how-to-update-joined-tables-using.html)

The first solution gives me the following code:
`				$this->db->set('attribute_values.attribute_datetime','attribute_values.attribute_value');
				$this->db->set('attribute_value', NULL);
				$this->db->where('attribute_links.definition_id',$definition_id);
				$success = $this->db->update('attribute_values JOIN attribute_links ON attribute_values.attribute_id = attribute_links.attribute_id');
`
But this generates the following query:
`
UPDATE 'attribute_values JOIN attribute_links ON attribute_values'.'ospos_attribute_id =' 'attribute_links'.'attribute_id' SET 'ospos_attribute_values'.'attribute_datetime' = 'attribute_values.attribute_value', 'attribute_value' = NULL
WHERE 'ospos_attribute_links'.'definition_id' = '8'
`

You can see that the format is correct but it's inserting "ospos_" on attribute_id and not inserting "ospos_" in other areas.  If I hardcode in the "ospos_" reference it still appends it onto the attribute_id field.  Is the feature that appends ospos_ our design or is that something that query builder does?  Any idea how to turn it off or correct it so that it gets inserted in the correct place?

**Comment by objecttothis** [comment]
I submitted an issue on the CI github, because it seems that this is an issue with Query Builder.  I wouldn't think it would be unreasonable to perform a join with an update.

Do you know of any places in the code where we are updating fields in a join?  Am I just going to need to perform two queries?  One query to pull all attribute_id's from attribute_links and then use that array to build an associative array and update attribute_values with that associative array in the second query?

**Comment by jekkos** [comment]
Ok if possible I'd split it in two queries for now. And then maybe wait if someone replies on the CI github issue? Also would expect this to work.

**Comment by objecttothis** [comment]
Yes, I will do that.

**Comment by objecttothis** [comment]
OK, so in the end I had to abandon Query Builder altogether and use db->query() and harcode the query.  Here is why:
I broke the query into two

> 				$this->db->select('attribute_id');
>				$this->db->from('attribute_links');
>				$this->db->where('definition_id',$definition_id);
>								
>				foreach($this->db->get()->result_array() as $row)
>				{
>					$attribute_ids[] = $row['attribute_id'];
>				}
>								
>				$this->db->set('attribute_datetime','attribute_value');
>				$this->db->set('attribute_value', NULL);
>				$this->db->where_in('attribute_id',$attribute_ids);
>				$success = $this->db->update('attribute_values');

The problem is that this generates an incredibly long array (in my case over 12,000 attribute_ids) and WHERE IN() for SQL was never meant to handle long arrays.  MYSQL on my box could handle it but the query alone took 30 seconds.  Query Builder broke altogether and just generated an error.  

When I ditched Query Builder altogether and wrote this code it worked perfectly: 

> 				$query = 'UPDATE ospos_attribute_values ';
>				$query .= 'INNER JOIN ospos_attribute_links ';
>				$query .= 'ON ospos_attribute_values.attribute_id = ospos_attribute_links.attribute_id ';
>				$query .= 'SET attribute_datetime = attribute_value, ';
>				$query .= 'attribute_value = NULL ';
>				$query .= 'WHERE definition_id = '.$definition_id;
>				$this->db->query($query);

The only other solution that I can think of is putting the query builder inside the foreach loop to set the rows one at a time, but that means 12,000+ queries.  Not a good idea.

The downside to my solution is that it hardcodes in the dbprefix because dbprefix is a Query Builder thing and it doesn't get appended onto query() calls.  This means that if someone manually changes dbprefix in the database.php file then this code breaks.  For this reason and others that I've outlined in #2361 I think we should deprecate the use of dbprefix and migrate the code/sql scripts/database to not include ospos_

**Comment by objecttothis** [comment]
I posted the question to the CI forum about this here (https://forum.codeigniter.com/thread-72820.html)

**Comment by objecttothis** [comment]
I created a pull request #2362. This enables conversion in the database for the following:

    * TEXT -> DROPDOWN
    * TEXT -> DATETIME
    * DROPDOWN -> TEXT


Javascript has been implemented in the form to prevent all other conversions. Currently DATETIME->TEXT is disabled because attribute_datetime does not have a UNIQUE constraint but attribute_value does. It could be implemented in the future but would require a check for each attribute_datetime to see if the attribute_value already exists and then change the attribute_id in the attribute_links table for that row. This is not a feature that we need so I'm not implementing it for now.
